### ISSUE TITLE
Configure Armeria to Return gRPC Status Code for Invalid Path

### ISSUE THREAD

**Issue by SeanWhoCodes** [line/armeria#4548]
## Overview

For an Armeria server that is built with minimal settings and has a simple `GrpcService`, the server returns only HTTP status code `404` but not any gRPC status such as `12` (UNIMPLEMENTED) when an invalid path is given.

Is there a way to configure the Armeria server so that it returns gRPC status code along with HTTP status?

## Examples

Below is an example server that uses `http4s-armeria` as backend:

```scala
import cats.effect.{ExitCode, IO, IOApp}
import org.http4s.armeria.server.ArmeriaServerBuilder

object Main extends IOApp {

  override def run(args: List[String]): IO[ExitCode] = {
    val grpcService = ??? // simple gRPC service defined somewhere else

    ArmeriaServerBuilder[IO]
      .bindHttp(9876, "0.0.0.0")
      .withHttpService(grpcService)
      .resource.useForever.as(ExitCode.Success)
  }
}
```

Example query command:

```shell
curl -v -H 'content-type: application/json' -d '{}' 'http://localhost:9876/io.test.NonExistentService/InvokeInvalidRpc'
```

In the output, only HTTP status `404` is available but there is no `grpc-status` in either the response headers or body.

## Use Case

For gRPC clients that stick to the protobuf service definitions strictly, it is unlikely for them to send requests with invalid paths.

However, for gRPC servers deployed to AWS and sitting behind an application load balancer, AWS performs health checks against targets at path `/AWS.ALB/healthcheck` and expects gRPC status `12` (UNIMPLEMENTED) by default if gRPC is used as the protocol version (see reference [here](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/target-group-health-checks.html#health-check-settings)). 

It'd be great if Armeria server could be configured to return gRPC status (such as `12`) for invalid paths.

**Comment by minwoox** [comment]
If you want to add the gRPC status to all requests that are sent to non-exist paths, we can use a decorator for the workaround
```java
sb.decorator((delegate, ctx, req) -> {
        final HttpResponse response = delegate.serve(ctx, req);
        if (!ctx.config().route().isFallback()) {
            return response;
        }
        return response.mapHeaders(...); 
    });
```
I think we can add an API something like `ServerBuilder.fallbackDecorator(...)` for such a case.

**Comment by ikhoon** [comment]
Hello, https://github.com/http4s/http4s-armeria user! 😆
`fallbackDecorator` would be a good workaround. But I also think we can use the default headers of
https://github.com/line/armeria/pull/4520 to set `grpc-status` 12 for `404 Not Found`. 
Note: A conditional default headers depending on response headers hasn't been implemented yet. I'd like to add the feature to the PR.
Users might set the default headers to `ServerBuilder`:
```java
Server.builder()
      .setHeader(HttpStatus.NOT_FOUND, "grpc-status", 12)
      // Or dynamically set depending on the status predicate
      .setHeader(status -> ..., "grpc-status", 12)
      // More than this flexibility, it would be better to use 
      // HTTP decorators
```

**Comment by ikhoon** [comment]
> AWS performs health checks against targets at path /AWS.ALB/healthcheck and expects gRPC status 12 (UNIMPLEMENTED) by default if gRPC is used as the protocol version (see reference [here](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/target-group-health-checks.html#health-check-settings)).

I've just checked the AWS health check reference. (http4s-)Armeria is able to handle both REST API and gRPC in a single port so you can simple add a health check service to `/AWS.ALB/healthcheck` to return `12` or configure the service as a normal HTTP/2 service to return `200 OK`.

**Comment by SeanWhoCodes** [comment]
> so you can simple add a health check service to /AWS.ALB/healthcheck to return 12 or configure the service as a normal HTTP/2 service to return 200 OK.

Really appreciate your responses team! 👍 Marking the gRPC status for a known rpc method as `UNIMPLEMENTED` sounds contrived and may be a bit confusing. Before the default header feature #4520 gets released and `fallbackDecorator` becomes an option, I may just stick with the decorator approach.

**Comment by SeanWhoCodes** [comment]
For reference, I implemented the 'not found' decorator as below that sets `grpc-status: 12` in both response headers and trailers. That should be convincing enough for the health checks on AWS load balancers.

Also, I hesitated but there seems to be a need to keep throwing exceptions that don't relate to 'not found'. It is not very idiomatic to throw exceptions in Scala, but I believe other decorators that wrap this fallback one may still need those exceptions. Please correct me if my understanding is incorrect. Thanks team!

```scala
package com.reagroup.psapi

import cats.effect.{ExitCode, IO, IOApp}
import com.linecorp.armeria.common.{HttpRequest, HttpResponse, HttpStatus}
import com.linecorp.armeria.server.{HttpService, HttpStatusException, ServiceRequestContext}
import org.http4s.armeria.server.ArmeriaServerBuilder

import java.util.Collections
import scala.util.Try

object ScratchMain extends IOApp {

  override def run(args: List[String]): IO[ExitCode] = {
    val grpcService = ??? // simple gRPC service defined somewhere else

    ArmeriaServerBuilder[IO]
      .bindHttp(9876, "0.0.0.0")
      .withHttpService(grpcService)
      .withDecorator(handleNotFound _)
      .resource.useForever.as(ExitCode.Success)
  }

  // 👇 Decorator method that handles 'not found'.
  private def handleNotFound(delegate: HttpService, ctx: ServiceRequestContext, req: HttpRequest): HttpResponse =
    Try(delegate.serve(ctx, req)).toEither.left.map {
      case httpStatusException: HttpStatusException
          if httpStatusException.httpStatus() == HttpStatus.NOT_FOUND =>
        val unimplementedStatus = Collections.singletonMap("grpc-status", "12").entrySet()
        HttpResponse.builder()
          .notFound()
          .headers(unimplementedStatus)
          .trailers(unimplementedStatus)
          .build()
      case throwable =>
        throw throwable
    }.merge
}
```

**Comment by ikhoon** [comment]
> I hesitated but there seems to be a need to keep throwing exceptions that don't relate to 'not found'.

Oops. I thought we removed all code that throws `HttpStatusException` or `HttpResponseException` apart from the default healthcheck service. This is because stream operators in `HttpResponse` can't be used correctly.
We need to fix `FallbackService` to return a normal `HttpResponse`. Before that, I recommend you use `ServerErrorHandler` as a workaround.
```java
sb.errorHandler(new ServerErrorHandler() {
    @Override
    public @Nullable HttpResponse onServiceException(ServiceRequestContext ctx, Throwable cause) {
        if (cause instanceof HttpStatusException) {
            if (((HttpStatusException) cause).httpStatus() == HttpStatus.NOT_FOUND) {
                return HttpResponse.builder()
                                   .notFound()
                                   .header("grpc-status", "12")
                                   .build();
            }
        }

        // Delegate to the default handler
        return null;
    }
});
```

Note that this workaround would no longer work if `FallbackService` returns a normal `HttpResponse`. Therefore, unit tests that ensure not-found response has `grpc-status 12` are necessary to detect the change.

**Comment by SeanWhoCodes** [comment]
> I recommend you use `ServerErrorHandler` as a workaround.

[ServerBuilder::errorHandler](https://javadoc.io/doc/com.linecorp.armeria/armeria-javadoc/latest/com/linecorp/armeria/server/ServerBuilder.html#errorHandler(com.linecorp.armeria.server.ServerErrorHandler)) does seem to be a better place to handle exceptional scenarios than a generic decorator where normal / valid requests also need to pass through. 👍 

> Note that this workaround would no longer work if `FallbackService` returns a normal `HttpResponse`. Therefore, unit tests that ensure not-found response has `grpc-status 12` are necessary to detect the change.

Thanks for the heads-up! What would be the recommended approach to address this in future when `FallbackService` gets refactored and no longer throws exceptions?

**Comment by ikhoon** [comment]
Sent a patch to fix `FallbackService`. #4552

> What would be the recommended approach to address this in future when FallbackService gets refactored and no longer throws exceptions?

A decorator could get you more freedom and fluent API in addition to handling 404 not found.
The default headers are slightly performant because it is optimized for headers but the difference would be small.

**Comment by ikhoon** [comment]
I tried to implement the feature that dynamically sets/adds the default headers depending on an HTTP status. But it was tricky to handle `addHeaders()`. If we use a function to store the default value, we don't know if a header is set by `ResponseHeaders` or the default headers until it is evaluated.

For that reason, please use the decorator pattern. The following pattern won't be implemented. 
```java
.setHeader(HttpStatus.NOT_FOUND, "grpc-status", 12)
// Or dynamically set depending on the status predicate
.setHeader(status -> ..., "grpc-status", 12)
```

**Comment by ikhoon** [comment]
Accidentally clicked the close button. 😅

**Comment by SeanWhoCodes** [comment]
> For that reason, please use the decorator pattern. The following pattern won't be implemented.

Thanks for the clarification! I'll leave a code snippet below for future reference.

```scala
object Main extends IOApp {

  private val grpcService = ??? // simple gRPC service defined somewhere else
  private val unimplementedStatus = Collections.singletonMap("grpc-status", "12")

  override def run(args: List[String]): IO[ExitCode] =
    ArmeriaServerBuilder[IO]
      .bindHttp(9876, "0.0.0.0")
      .withHttpService(grpcService)
      .withDecorator(handleNotFoundAfterFix _) // 👈
      .resource.useForever.as(ExitCode.Success)

  // 👇 Decorator that handles 'not found' after fix #4552 gets merged (>=v1.21.0)
  private def handleNotFoundAfterFix(delegate: HttpService, ctx: ServiceRequestContext, req: HttpRequest): HttpResponse = {
    val response = delegate.serve(ctx, req)
    if (ctx.config().route().isFallback) {
      response
        .mapHeaders(_.toBuilder.add(unimplementedStatus).build())
        .mapTrailers(_.toBuilder.add(unimplementedStatus).build())
    } else
      response
  }
}
```

**Comment by minwoox** [comment]
>  I'll leave a code snippet below for future reference.

That looks great. 😄
